[Review Ready] Positronic Brains!

GunHog opened the pull request:
Shamelessly rips off https://github.com/NTStation/NTstation13/pull/772 by @RemieRichards!

_This is a Work In Progress, the following things need to be done before completion:_
- [x] ~~Name and create an appropriate preference and job-ban flag for them, requested by @Ikarrus. I need feedback from admins on this one.~~ I had to put pAI and Posibrain in the same preference, due to the flag being full.
- [x] Draft a changelog for Jordie, in my own blood.
- [x] Perform extended tests on them, and fix any obvious bugs.
- [x] Perform a sample DB test on the admin related changes. <b>Special thanks to SoS for helping me with this!</b>

Positronic Brains (https://tgstation13.org/phpBB/viewtopic.php?f=9&t=3448) are special MMIs that accept ghosts. They are originally a port from Bay, PR'ed to NT by RR, and now yoinked for /tg/ by me!

Notable differences in my version:
- <b>New!</b> Borgs created with Xeno brains are now called "Xenoborgs".
- Positronic Brains will not have binary or a radio. They will hear and speak only locally.
- They are selected in a similar manner as a pAI. A human player picks it up and activates it.
- They are informed that they have forgotten their past lives.
- They are functionally the same as AIs and Cyborgs. There is only one difference from normal MMIs: They will be named "Android" instead of "Cyborg".
- As mech pilots, they cannot be antagonist, and are informed to do no harm. _subject to change based on feedback_
- Cyborgs created during the AI Malfunction gamemode gain Malfunction laws even if they are configured to be unslaved. The same will go for Androids, as they are not meant to be different.

<b>Ban panel changes: Admins, please note:</b>
- There are now individual job-ban options for pAI, Drones, and Positronic Brains, "Ghost Roles" category on the job-ban panel.
- Deathsquad moved from Antagonist section to "Ghost Roles"; Emergency Response Team are included in Deathsquad bans! 

Androids created with Positronic brains:
![screenshot 2015-04-24 09 33 17](https://cloud.githubusercontent.com/assets/4955510/7320957/28698fb2-ea65-11e4-8d6b-8e955b62becb.png)

Empty brain and Android:
![screenshot 2015-04-24 09 33 19](https://cloud.githubusercontent.com/assets/4955510/7320970/353a2e68-ea65-11e4-89cb-65d1be60d06e.png)

Commit:
Positronic Brains!

-Adds positronic brains
- Fixes a runtime in mechs for pilots without DNA

Miauw62 commented:
this lets you build huge borg armies unless you make these things very expensive, and i've already heard admins complain about drones being unaccountable griff machines.

RemieRichards commented on code/modules/mob/living/carbon/brain/posibrain.dm:
```diff
@@ -0,0 +1,162 @@
+	usr << msg
+	return
+
+/obj/item/device/mmi/posibrain/emp_act(severity)
```
inherited already. Remove it all.

RemieRichards commented on code/game/mecha/mecha.dm:
```diff
@@ -1137,6 +1137,9 @@ obj/mecha/proc/can_use(mob/user)
 			if(mmi.brainmob)
 				occupant.loc = mmi
 			mmi.mecha = null
+			if(istype(mmi,/obj/item/device/mmi/posibrain))
```
You can accomplish this by giving MMIs an ```update_icon()``` and overriding it on Posibrains.
(MMIs handle their icons in attackby, also with an ```istype()``` for alien brains, so that needs a cleanup regardless)

RemieRichards commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -123,6 +124,9 @@
 		if(T)	mmi.loc = T
 		if(mmi.brainmob)
 			mind.transfer_to(mmi.brainmob)
+			if(istype(mmi,/obj/item/device/mmi/posibrain))
```
see previous point on updating icons.

RemieRichards commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -236,7 +240,7 @@
 		changed_name = custom_name
 	else
-		changed_name = "[(designation ? "[designation] " : "")]Cyborg-[num2text(ident)]"
+		changed_name = "[(designation ? "[designation] " : "")][istype(mmi, /obj/item/device/mmi/posibrain) ? "Android" : "Cyborg"]-[num2text(ident)]"
```
just give MMIs a designation variable, that defaults to "Cyborg" and is set to "Android" on Posibrains.

RemieRichards commented on code/modules/mob/living/carbon/brain/posibrain.dm:
```diff
@@ -0,0 +1,162 @@
+/obj/item/device/mmi/posibrain/New()
+
+	brainmob = new(src)
+	brainmob.name = "[pick(list("PBU","HIU","SINA","ARMA","OSI","HBL","MSO"))]-[rand(100, 999)]"
```
>sneaking in references to HBL and MSO

>Not sneaking in one for me

Have you no shame?

GunHog replied:
No, Ma'am, I do not!

RemieRichards commented on code/modules/mob/living/carbon/brain/posibrain.dm:
```diff
@@ -0,0 +1,162 @@
+	brainmob.mind.assigned_role = "Positronic Brain"
+
+	var/turf/T = get_turf()
+	for (var/mob/M in viewers(T))
```
Baycode.

Swap it for visible_message()

RemieRichards commented on code/modules/mob/living/carbon/brain/posibrain.dm:
```diff
@@ -0,0 +1,162 @@
+	handle_posibrain_icon()
+
+	var/turf/T = get_turf()
+	for (var/mob/M in viewers(T))
```
visible_message()

ike709 commented:
heil NT

:+1:

dalmationer commented:
If you nerds want sprites i totally have some old ones you can use. Check it. http://imgur.com/O20bE6S
There's also some other shit on there like mobile PAIs but whatever. If you want it i can send u guys XCFs or whatevs.

pudl commented:
that it, /tg/ is dead. rip.

Commit:
Tweaks and changes!

GunHog commented:
More stuff added, Borgs now can be "Androids", "Cyborgs", or "Xenoborgs"! Added RR!

Amnestik commented:
>Positronic Brains will not have binary or a radio. They will hear and speak only locally.

How come? It'll make it difficult for them to follow their laws, prevent human harm and obey the AI.

GunHog commented:
>How come? It'll make it difficult for them to follow their laws, prevent human harm and obey the AI.

They are MMIs, and are meant to have the same limits as one. They will gain Binary if they are made into an android or AI. Mechs have built in radios, so they will have communication abilities if you make them a pilot. They cannot do anything but talk in brain form, anyway.

Amnestik commented:
oh, i see

On 27 April 2015 at 11:16, GunHog <notifications@github.com> wrote:

> How come? It'll make it difficult for them to follow their laws, prevent
> human harm and obey the AI.
>
> They are MMIs, and are meant to have the same limits as one. They will
> gain Binary if they are made into an android or AI. Mechs have built in
> radios, so they will have communication abilities if you make them a pilot.
> They cannot do anything but talk in brain form, anyway.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/tgstation/-tg-station/pull/9147#issuecomment-96447589>
> .
>

Commit:
pAI -> Ghost Roles

Changes the BE_PAI and "pAI" flag names to "BE_GHOST_ROLE" and "Ghost
Roles".
Name subject to change.

GunHog commented:
@Ikarrus, I changed the pAI flag to Ghost Roles, as asked. Basic ban-application testing seems to work. If anyone here does not like the new name, <b>Please supply a better one.</b>

Commit:
Changelog

Adds a changelog, drawn in blood, and performed according to the rites
as scribed in the Unholy One's guide to changelogs. Long may he reign.

PKPenguin321 commented:
If you're testing for bugs, can you make sure that a cyborg built with a positronic brain can't be deconstructed to yield a human brain? Because once this goes live that'll be one of the first things I'll be testing for, as it could be a way to make ghosts into humans.

GunHog commented:
That is working normally. Deconned and blown Androids drop their Posibrain as expected. I will hunt for any mistakes.

Ikarrus commented:
Am I correct in assuming the ban database stores the index instead of the string? So people who were previously pAI jobbanned aren't going to suddenly find themselves unbanned.

GunHog commented:
Oh.....god....dang it....Ikky, I do not know at all. I have no clue how the database works at all. I will need an admin-coder to comment. If this would unban people, I will revert the name change, and only change appearance on the panel to tell the admin it includes drones and posibrains.

MrStonedOne commented:
Ok, so the job ban name is used. not an index.

That being said. This is one update query to fix.

````
UPDATE SS13_bans SET JOB = "Ghost Roles" WHERE JOB = "pAI"
````
(just make a change log .sql file with that in it and add info the sql change log about the change)

But, a better idea would be to make this a new category in the job ban panel. so we can ban from each 3 things (drones, pai, posbrain) separately, or hit the category button to ban from all 3.

MrStonedOne commented on code/modules/mob/living/simple_animal/friendly/drone/drones_as_items.dm:
```diff
@@ -16,7 +16,7 @@
 
 /obj/item/drone_shell/attack_ghost(mob/user)
-	if(jobban_isbanned(user,"pAI"))
+	if(jobban_isbanned(user,"Ghost Roles"))
```
replace this with drone.

MrStonedOne commented on code/modules/mob/living/carbon/brain/posibrain.dm:
```diff
@@ -0,0 +1,146 @@
+
+/obj/item/device/mmi/posibrain/proc/request_player()
+	for(var/mob/dead/observer/O in player_list)
+		if(jobban_isbanned(O, "Ghost Roles"))
```
replace this with posibrain

MrStonedOne commented on code/controllers/subsystem/pai.dm:
```diff
@@ -181,7 +181,7 @@ var/datum/subsystem/pai/SSpai
 /datum/subsystem/pai/proc/requestRecruits()
 	for(var/mob/dead/observer/O in player_list)
-		if(jobban_isbanned(O, "pAI"))
+		if(jobban_isbanned(O, "Ghost Roles"))
```
pop this back to pAI

Dorsisdwarf commented:
I don't see why we'd need the ability to ban individual ghost roles.

As far as I'm concerned, someone who's being a shitlord/metagamer as a drone is going to be the same as a pAI, and definitely a posibrain. Even if they aren't,  bans are applied across the type of job. Beat a prisoner to death for fun as a warden and you get a sec jobban, not a warden jobban

Commit:
Ban panel  and preferences changes

- Deathsquad moved from Antagonist to Ghost Roles.
- pAI jobbans split into drone, pAI, and posibrains.
- Posibrains back under pAI preference flag (the byte is full)
- Fixed Posibrain sprite if it fails to get a ghost.

GunHog commented:
>I don't see why we'd need the ability to ban individual ghost roles.

>As far as I'm concerned, someone who's being a shitlord/metagamer as a drone is going to be the same as a pAI, and definitely a posibrain. Even if they aren't, bans are applied across the type of job. Beat a prisoner to death for fun as a warden and you get a sec jobban, not a warden jobban

You can now do both. If you want to jobban someone from all ghost roles, you may do so, but some admins might think differently. I am giving them both choices, because admins need the power!

Commit comments:
I moved Deaf Squids to ghost roles, per MSO's request.
I split the job-bans for drone, pAI, and Posibrains.
The stupid byteflag is full, so I had to shove pAI and Posibrains back together in preferences.

<b>PLEASE LET THIS BE THE LAST I HAVE TO MESS WITH ADMIN STUFF. It is scary, and I do not want admins angry with me!</b>

GunHog commented:
I cannot find anything wrong with it now, ready for maintainer review!
